### PR TITLE
Added functionality for forwards & retrieve longest prefix

### PR DESCRIPTION
**Pull request by alinabuzila**
- added forward request functionality
- added get longest prefix functionality
- TODO: ask Dave if he implemented the arp handler and the get longest prefix does not acc consider longest it just compares by masking
- TODO: fix warnings

### CODE REVIEW

File: cs144_lab3/router/sr_router.c [hunk 1]
```diff
@@ -296,6 +299,59 @@ void sr_forward_IP_packet(struct sr_instance *sr, sr_if_t *interface, unsigned i
...
+    /* otherwise we found the associated MAC addy of next hop IP addy */
+    printf("Next hop MAC found in arp cache\n");
+    ether_hdr = (sr_ethernet_hdr_t*) packet;
+
+    unsigned char *ether_src = sr_get_interface(sr, longest_prefix_entry->interface)->addr;
+    /* ether_hdr->ether_shost = ether_src */
+    memcpy(ether_hdr->ether_shost, ether_src, ETHER_ADDR_LEN);
+    memcpy(ether_hdr->ether_dhost, next_hop_arp_entry->mac, ETHER_ADDR_LEN);
+
+    sr_send_packet(sr, packet, len, longest_prefix_entry->interface);
```
Comment (davepetrov) [inline cs144_lab3/router/sr_router.c:334]: make sure to free next_hop_arp_entry pointer.

The function u use to call it says explicitly that "You must free the returned structure if it is not NULL. " cus it mallocs the pointer in the sr_arpcache_lookup() funtion

...dont want leeks

File: cs144_lab3/router/sr_router.c [hunk 2]
```diff
@@ -296,6 +299,59 @@ void sr_forward_IP_packet(struct sr_instance *sr, sr_if_t *interface, unsigned i
...
+
+  /* if no entry in table matches dest */
+  if (longest_prefix_entry == NULL) {
+    printf("Destination net unreachable. Sending ICMP reply\n");
+    sr_send_ICMP_reply(sr, ether_hdr, 3, 0, packet, interface, len);
+  }
+  else {
+    /* if an antry matched */
+    uint32_t next_hop_ip = longest_prefix_entry->gw.s_addr;
+    fprintf(stderr, "...... the match was ......... ");
```
Comment (davepetrov) [inline cs144_lab3/router/sr_router.c:313]: wouldnt call this a stderr msg. Simple debug msg  u wanna send to console.

File: cs144_lab3/router/sr_router.c [hunk 3]
```diff
@@ -296,6 +299,59 @@ void sr_forward_IP_packet(struct sr_instance *sr, sr_if_t *interface, unsigned i
...
+  struct sr_rt *longest_match_entry = NULL;
+  struct sr_rt* entry;
+
+  /* iteratre through entries routing table */
+  for (entry = routing_table_entry; entry != NULL; entry = entry->next) {
+    /* if we get a match when applying the mask */
+    /* in a2 we applied subnet mask to destination and had subnetNum col 
+    subnetNum col means the subnet mask was applied to the destination IP addy */
+    if ((ip_hdr->ip_dst & entry->mask.s_addr) == (entry->dest.s_addr & entry->mask.s_addr)) {
+      longest_match_entry = entry;
```
Comment (davepetrov) [inline cs144_lab3/router/sr_router.c:350]: we still gotta figure out how to do the longest prefix matching. another pr for this feature or should we keep in this pr?

File: cs144_lab3/router/sr_router.c [hunk 4]
```diff
@@ -42,6 +44,8 @@ void set_IP_header(sr_ethernet_hdr_t *ether_hdr,
   sr_if_t *interface,
   uint8_t type, uint8_t code);
 void set_ICMP_header(sr_ip_hdr_t *ip_hdr, uint8_t type, uint8_t code, uint8_t *new_packet, uint8_t *packet);
```
Comment (kingpreyansh) [inline cs144_lab3/router/sr_router.c:46]: Add all of these functions to the sr_router.h file instead of here

Comment (alinabuzila) [reply cs144_lab3/router/sr_router.c:46]: Done thanks

**Review by davepetrov — CHANGES_REQUESTED** [review]

**Pull request merged**